ZeroK-RTS/Zero-K-Infrastructure
# Simple implementation of MM map veto

Addresses https://github.com/ZeroK-RTS/Zero-K-Infrastructure/issues/2444. I have experience in other web frameworks but no previous experience in C#, ASP or this specific codebase, so please check for stupid errors or code improvements I'm not aware of.

The first commit contains all MVC architecture, and the second commit is only the changes to the matchmaker itself. The second commit is the one most likely to cause problems, so it can be reverted on its own.

Testing I've done locally:
- Verify multiple players can store their bans
- Verify bad input doesn't cause any problems
- Verify map bans are applied in a 1v1 scenario
- Verify setting map bans to 0 for 1v1 results in the same behavior as before my changes
- Verify bans for previous matchmaking maps are ignored
- Verify accidentally banning all the maps doesn't break matchmaking
- Verify map autocomplete still works in Battle list and elsewhere

I cannot verify actual hosting of a game locally due to insufficient memory on my machine, or more than 1v1. I don't believe there's a working test instance for matchmaking at present, so it's probably worth having someone else sanity check it locally if they can.

Room for future improvement: have separate bans for 1v1 and team games and a less spartan UI. Right now applying bans to maps only used in team games requires "losing" bans for 1v1, and selecting maps is a little confusing since all matchmaking games are returned and it's not clear which ones are only used in 1v1. This is probably fine for now since team game MM is pretty niche, and it can be worked around (by increasing the total number of bans per user or just having users editing their bans to be more targeted at team games).
